File: app/src/main/java/com/dmrandevu/gallery/media/censor/BeepPatcher.kt
package com.dmrandevu.gallery.media.censor

import androidx.media3.common.util.UnstableApi
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

/**
 * Renders the short stretches of audio that will replace the swearing.
 *
 * Only the censor windows are touched, and only they are separated — the expensive part runs on a
 * second or two of audio per swear word rather than the whole clip. Separating a 35 second video
 * end to end took 84 seconds of desktop CPU in the spike, which on a phone would be the whole
 * export; this way it is proportional to how much swearing there is.
 *
 * Everything outside a window is never read, never re-encoded here, and comes through the export
 * untouched.
 */
@UnstableApi
object BeepPatcher {

    /** A stretch of replacement audio, at the source's own rate and channel count. */
    class PcmPatch(val startFrame: Long, val samples: ShortArray, val channelCount: Int) {
        val frameCount: Long get() = samples.size.toLong() / channelCount
        val endFrame: Long get() = startFrame + frameCount
    }

    /**
     * Builds one patch per window in [windows].
     *
     * [onProgress] reports 0..1 across all of them.
     */
    suspend fun render(
        audio: AudioTrackDecoder.DecodedAudio,
        windows: List<CensorWindow>,
        separator: VocalSeparator,
        onProgress: (Float) -> Unit
    ): List<PcmPatch> {
        if (windows.isEmpty()) return emptyList()
        val patches = ArrayList<PcmPatch>(windows.size)
        for ((index, window) in windows.withIndex()) {
            currentCoroutineContext().ensureActive()
            patches.add(render(audio, window, separator))
            onProgress((index + 1f) / windows.size)
        }
        return patches
    }

    private suspend fun render(
        audio: AudioTrackDecoder.DecodedAudio,
        window: CensorWindow,
        separator: VocalSeparator
    ): PcmPatch {
        val rate = audio.sampleRate
        val channels = audio.channelCount

        // The patch covers the window plus a crossfade at each end, so it can be eased into the
        // untouched audio around it rather than butted against it.
        val fadeFrames = rate * PcmOps.CROSSFADE_MS / 1000
        val windowStart = (window.startUs * rate / 1_000_000L)
        val windowEnd = (window.endUs * rate / 1_000_000L)
        val patchStart = (windowStart - fadeFrames).coerceAtLeast(0L)
        val patchEnd = (windowEnd + fadeFrames).coerceAtMost(audio.frameCount)
        val patchFrames = (patchEnd - patchStart).toInt()
        if (patchFrames <= 0) {
            return PcmPatch(patchStart, ShortArray(0), channels)
        }

        // Separation works at its own rate in stereo, so the slice is converted, separated, and
        // converted back. Context on either side is real audio wherever there is any: the model
        // reads a block at a time, and feeding it silence at the edges makes it hear the join.
        val original = slice(audio, patchStart, patchFrames)
        val background = separateBackground(original, rate, channels, separator)

        PcmOps.mixBeepInto(
            destination = background,
            startFrame = (windowStart - patchStart).toInt(),
            endFrame = (windowEnd - patchStart).toInt(),
            sampleRate = rate
        )
        PcmOps.crossfadeEdges(original, background, fadeFrames)

        return PcmPatch(patchStart, PcmOps.interleave(background), channels)
    }

    /**
     * Takes the voice out of [original], leaving the background.
     *
     * The slice is padded out to whole separation blocks with the audio that really surrounds it,
     * and the padding is discarded afterwards.
     */
    private suspend fun separateBackground(
        original: Array<FloatArray>,
        rate: Int,
        channels: Int,
        separator: VocalSeparator
    ): Array<FloatArray> {
        val frames = original[0].size
        val atModelRate = resampleChannels(original, rate, PcmOps.SEPARATION_SAMPLE_RATE)
        val stereo = toStereo(atModelRate)
        val modelFrames = stereo[0].size

        val separated = Array(2) { FloatArray(modelFrames) }
        var offset = 0
        while (offset < modelFrames) {
            currentCoroutineContext().ensureActive()
            // Each block is read from the real signal, centred so the trimmed edges fall outside
            // the part being kept.
            val blockStart = offset - Stft.TRIM
            val block = Array(2) { channel ->
                FloatArray(Stft.CHUNK) { i ->
                    val at = blockStart + i
                    if (at in 0 until modelFrames) stereo[channel][at] else 0f
                }
            }
            val voice = separator.vocalsIn(block)
            val take = minOf(Stft.USABLE, modelFrames - offset)
            for (channel in 0..1) {
                for (i in 0 until take) {
                    val source = Stft.TRIM + i
                    separated[channel][offset + i] =
                        stereo[channel][offset + i] - voice[channel][source]
                }
            }
            offset += Stft.USABLE
        }

        val backAtSourceRate = resampleChannels(
            separated, PcmOps.SEPARATION_SAMPLE_RATE, rate
        )
        // Resampling rarely lands on exactly the frame count it started from; the patch has to be
        // the length the caller expects, so it is trimmed or held at its last value.
        return Array(channels) { channel ->
            val source = backAtSourceRate[channel.coerceAtMost(backAtSourceRate.size - 1)]
            FloatArray(frames) { i -> if (i < source.size) source[i] else 0f }
        }
    }

    private fun slice(
        audio: AudioTrackDecoder.DecodedAudio,
        startFrame: Long,
        frames: Int
    ): Array<FloatArray> {
        val channels = audio.channelCount
        return Array(channels) { channel ->
            FloatArray(frames) { i ->
                val index = ((startFrame + i) * channels + channel).toInt()
                if (index in audio.samples.indices) audio.samples[index] / 32768f else 0f
            }
        }
    }

    private fun toStereo(channels: Array<FloatArray>): Array<FloatArray> = when (channels.size) {
        2 -> channels
        // A mono track is fed to both sides; the model has never seen anything else.
        1 -> arrayOf(channels[0], channels[0].copyOf())
        else -> arrayOf(channels[0], channels[1])
    }

    private fun resampleChannels(
        channels: Array<FloatArray>,
        fromRate: Int,
        toRate: Int
    ): Array<FloatArray> {
        if (fromRate == toRate) return channels
        return Array(channels.size) { channel ->
            val shorts = ShortArray(channels[channel].size) { i ->
                (channels[channel][i] * 32768f).coerceIn(-32768f, 32767f).toInt().toShort()
            }
            val resampled = PcmOps.resample(shorts, 1, fromRate, toRate)
            FloatArray(resampled.size) { resampled[it] / 32768f }
        }
    }
}

File: ios/DMRandevuGaleri/Media/Censor/BeepPatcher.swift
import Foundation

/// A stretch of replacement audio, at the source's own rate and channel count.
struct PcmPatch {
    var startFrame: Int64
    var samples: [Int16]
    var channelCount: Int

    var frameCount: Int64 { Int64(samples.count) / Int64(channelCount) }
    var endFrame: Int64 { startFrame + frameCount }
}

/// Renders the short stretches of audio that will replace the swearing.
///
/// Only the censor windows are touched, and only they are separated — the expensive part runs on a
/// second or two of audio per swear word rather than the whole clip. Separating a 35 second video
/// end to end took 84 seconds of desktop CPU when this was measured; this way it is proportional
/// to how much swearing there is.
enum BeepPatcher {

    /// Builds one patch per window. `onProgress` reports 0…1 across all of them.
    static func render(
        audio: AudioTrackDecoder.DecodedAudio,
        windows: [CensorWindow],
        separator: VoiceSeparator,
        onProgress: (Double) -> Void = { _ in }
    ) throws -> [PcmPatch] {
        guard !windows.isEmpty else { return [] }
        var patches: [PcmPatch] = []
        for (index, window) in windows.enumerated() {
            try Task.checkCancellation()
            patches.append(try render(audio: audio, window: window, separator: separator))
            onProgress(Double(index + 1) / Double(windows.count))
        }
        return patches
    }

    private static func render(
        audio: AudioTrackDecoder.DecodedAudio,
        window: CensorWindow,
        separator: VoiceSeparator
    ) throws -> PcmPatch {
        let rate = audio.sampleRate
        let channels = audio.channelCount

        // The patch covers the window plus a crossfade at each end, so it can be eased into the
        // untouched audio around it rather than butted against it.
        let fadeFrames = Int64(rate * PcmOps.crossfadeMs / 1000)
        let windowStart = window.startUs * Int64(rate) / 1_000_000
        let windowEnd = window.endUs * Int64(rate) / 1_000_000
        let patchStart = max(0, windowStart - fadeFrames)
        let patchEnd = min(audio.frameCount, windowEnd + fadeFrames)
        let patchFrames = Int(patchEnd - patchStart)
        guard patchFrames > 0 else {
            return PcmPatch(startFrame: patchStart, samples: [], channelCount: channels)
        }

        let original = slice(audio, from: patchStart, frames: patchFrames)
        var background = try separateBackground(
            original, rate: rate, channels: channels, separator: separator
        )

        PcmOps.mixBeepInto(
            &background,
            startFrame: Int(windowStart - patchStart),
            endFrame: Int(windowEnd - patchStart),
            sampleRate: rate
        )
        PcmOps.crossfadeEdges(original: original, patched: &background, frames: Int(fadeFrames))

        return PcmPatch(
            startFrame: patchStart,
            samples: PcmOps.interleave(background),
            channelCount: channels
        )
    }

    /// Takes the voice out, leaving the background.
    ///
    /// The slice is padded out to whole separation blocks with the audio that really surrounds it,
    /// and the padding is discarded afterwards — feeding the model silence at the block edges
    /// makes it hear the join.
    private static func separateBackground(
        _ original: [[Float]],
        rate: Int,
        channels: Int,
        separator: VoiceSeparator
    ) throws -> [[Float]] {
        let frames = original[0].count
        let atModelRate = try resampleChannels(
            original, from: rate, to: PcmOps.separationSampleRate
        )
        let stereo = toStereo(atModelRate)
        let modelFrames = stereo[0].count

        var separated = [[Float]](repeating: [Float](repeating: 0, count: modelFrames), count: 2)
        var offset = 0
        while offset < modelFrames {
            try Task.checkCancellation()
            // Each block is read from the real signal, centred so the trimmed edges fall outside
            // the part being kept.
            let blockStart = offset - Stft.trim
            let block = (0..<2).map { channel in
                (0..<Stft.chunk).map { i -> Float in
                    let at = blockStart + i
                    return at >= 0 && at < modelFrames ? stereo[channel][at] : 0
                }
            }
            let voice = try separator.vocals(in: block)
            let take = min(Stft.usable, modelFrames - offset)
            for channel in 0..<2 {
                for i in 0..<take {
                    separated[channel][offset + i] =
                        stereo[channel][offset + i] - voice[channel][Stft.trim + i]
                }
            }
            offset += Stft.usable
        }

        let back = try resampleChannels(separated, from: PcmOps.separationSampleRate, to: rate)
        // Resampling rarely lands on exactly the frame count it started from; the patch has to be
        // the length the caller expects.
        return (0..<channels).map { channel in
            let source = back[min(channel, back.count - 1)]
            return (0..<frames).map { $0 < source.count ? source[$0] : 0 }
        }
    }

    private static func slice(
        _ audio: AudioTrackDecoder.DecodedAudio,
        from startFrame: Int64,
        frames: Int
    ) -> [[Float]] {
        let channels = audio.channelCount
        return (0..<channels).map { channel in
            (0..<frames).map { i -> Float in
                let index = Int((startFrame + Int64(i)) * Int64(channels)) + channel
                return index >= 0 && index < audio.samples.count
                    ? Float(audio.samples[index]) / 32768
                    : 0
            }
        }
    }

    /// A mono track is fed to both sides; the model has never seen anything else.
    private static func toStereo(_ channels: [[Float]]) -> [[Float]] {
        switch channels.count {
        case 2: return channels
        case 1: return [channels[0], channels[0]]
        default: return [channels[0], channels[1]]
        }
    }

    private static func resampleChannels(
        _ channels: [[Float]],
        from fromRate: Int,
        to toRate: Int
    ) throws -> [[Float]] {
        if fromRate == toRate { return channels }
        return try channels.map { channel in
            let shorts = channel.map { Int16(max(-32768, min(32767, ($0 * 32768).rounded()))) }
            let resampled = try PcmOps.resample(
                shorts, channelCount: 1, from: fromRate, to: toRate
            )
            return resampled.map { Float($0) / 32768 }
        }
    }
}
